satpy/readers/netcdf_utils.py

```diff
@@ -287,3 +320,24 @@ def get(self, item, default=None):
             return self[item]
         else:
             return default
+
+    def get_and_cache_npxr(self, var_name):
+        """Get item as numpy-xarray and keep in cache."""
+        if var_name in self.cached_file_content:
+            return self.cached_file_content[var_name]
```

Comment:
Instead of implementing the cache ourselves here, could we use lru_cache with the right size around this function for example?

Reply:
So something like this?
```python
    @lru_cache
    def get_and_cache_npxr(self, var_name):
        """Get item as numpy-xarray and keep in cache."""
        v = self.file_content[var_name]
        if isinstance(v, xr.DataArray):
            return v
        return xr.DataArray(
            v[:], dims=v.dimensions, attrs=v.__dict__, name=v.name)
```

Reply:
Or `@cache`, as we won't be collecting that many variables, and the `functools` documentation says `@cache` is faster than `@lru_cache` because it doesn't need to handle deleting any excess items.

Reply:
Hmm, should also `collect_cache_vars()` be restructured to use `@cache`?

Reply:
I got this complaint from `flake8`: `B019 Use of `functools.lru_cache` or `functools.cache` on methods can lead to memory leaks. The cache may retain instance references, preventing garbage collection.` Should we be worried about this, or add `# noqa`?

Reply:
The docs for `@functools.cache` say "New in version 3.9.". I suppose we can't add such a new feature as long as we support 3.8? Or would we need some sort of fallback mechanism to `@lru_cache`?

Reply:
As you said, the amount of caching done here is limited, and will not eg leave files open, so I think we're good with a `# noqa`

Reply:
> The docs for `@functools.cache` say "New in version 3.9.". I suppose we can't add such a new feature as long as we support 3.8? Or would we need some sort of fallback mechanism to `@lru_cache`?

Oh, thanks, didn't notice! I'll add a fallback.

Reply:
would be actually interesting to know how much overhead we add for the caching... here I activated it for all small variables as a first tentative, but some of the variables are actually unlikely to ever be used again (e.g. calibration parameters after we have loaded a channel). I guess we're probably talking about microseconds, especially after we implement the read-only-at-the-first-chunk mechanism.. But I can still run some tests after the `@cache` implementation is here

Reply:
This now uses the caching from `functools`. I think changing the `collect_cache_vars()` to do the same is too big a change touching multiple readers that it has to be done in a separate PR, if at all.

Reply:
Please don't intentionally introduce memory leaks. Can we put this in a cached function instead of a method? And if you're adding workarounds to use the newer cache decorator, I think there is a compat.py module that might include one already.

Reply:
The caching is now in a separate function, and the `cache`/`lru_cache` moved to `satpy._compat` module.